### PR TITLE
finish first draft of project section, completing first draft of portfolio, complete first 3 parts of week 3 tutorial

### PR DESCRIPTION
**Pull request by ntarn**
still have some minor todo's left in with comments, but mostly just need to update links and descriptions

### CODE REVIEW

**Comment by ntarn** [comment]
I added the parts of the week 3 tutorial to the same branch because I thought they were just going to be small, easy "test" commits, but turns out they took more time and aren't necessarily part of the project-section. I'm going to create separate branch for the next few commits, so apologies for the prior 7 being in this branch!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -19,14 +19,35 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
+import com.google.gson.Gson;
+
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:25]: Remove extra new line.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -19,14 +19,35 @@
...
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:27]: Just some convention:

1. The first sentence should say: "A servlet..."
2. End all comments with "." to keep them consistent
3. TODO should come on the next line.
4. Another rule we follow here about TODOs is assigning them to yourself with your ldap, so this would look something like this:

```suggestion
/*
 * Servlet that returns some example content. 
 * TODO(ntarn): Modify this file to handle comments data.
 */
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -19,14 +19,35 @@
...
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  ArrayList<String> hardcodeMessages = new ArrayList<String>(Arrays.asList("Buenos Aires", "Cordoba", "La Plata"));
+	
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    // Send the JSON as the response
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:35]: End all comments with a "."

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -19,14 +19,35 @@
...
+    String json = convertToJsonUsingGson(hardcodeMessages);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+    // Send the html as the response
+    // response.setContentType("text/html;");
+    // response.getWriter().println("<h1>Hello Natalie!</h1>");
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:41]: Leftover code?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -19,14 +19,35 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import com.google.gson.Gson;
+
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  ArrayList<String> hardcodeMessages = new ArrayList<String>(Arrays.asList("Buenos Aires", "Cordoba", "La Plata"));
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:31]: We prefer to use arrays here.
I'd use the style guide: https://google.github.io/styleguide/jsguide.html#features-array-literals

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -19,14 +19,35 @@
...
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+    // Send the html as the response
+    // response.setContentType("text/html;");
+    // response.getWriter().println("<h1>Hello Natalie!</h1>");
+  }
+
+  /**
+   * Converts a ServerStats instance into a JSON string using the Gson library. Note: We first added
+   * the Gson library dependency to pom.xml.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:46]: Is this something you did?

File: portfolio/src/main/webapp/css/customstyles.css [hunk 7]
```diff
@@ -86,3 +97,25 @@ h1{
...
+  background: rgba(255, 255, 255, 0.1);
+}
+
+.resizemobiletesting {
+  width: 95%;
+  height: 20%;
+  background: rgba(255, 255, 255, 0.1);
+}
+
+/* the extended projects section used to be a contact section so actually fixing extended projects */
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/css/customstyles.css:113]: "The", and end with a "."

File: portfolio/src/main/webapp/css/customstyles.css [hunk 8]
```diff
@@ -86,3 +97,25 @@ h1{
...
+#fixcontactmargin {
+  padding-top: 4rem;
+  padding-bottom: 4rem;
+}
+
+.resizenlp{
+  width:95%;
+}
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/css/customstyles.css:121]:
```suggestion
.resizenlp {
  width: 95%;
}
```

File: portfolio/src/main/webapp/css/customstyles.css [hunk 9]
```diff
@@ -86,3 +97,25 @@ h1{
 .project-icons a:active {
   color: rgba(255, 255, 255, 1);
 }
+
+.resizewebappimg {
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/css/customstyles.css:101]: Use the naming convention: word1-word2-word3 as you did above.

Same for all the class and id names

File: portfolio/src/main/webapp/css/customstyles.css [hunk 10]
```diff
@@ -86,3 +97,25 @@ h1{
...
+}
+
+.resizemobiletesting {
+  width: 95%;
+  height: 20%;
+  background: rgba(255, 255, 255, 0.1);
+}
+
+/* the extended projects section used to be a contact section so actually fixing extended projects */
+#fixcontactmargin {
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/css/customstyles.css:114]: I'd rename this to something more descriptive. "fix" is quite ambiguous.

Instead, just call it "contact-margin" and add a comment describing what you're doing

File: portfolio/src/main/webapp/css/customstyles.css (hunk 9 above)
Comment (hayleejane3) [inline portfolio/src/main/webapp/css/customstyles.css:101]: The name is a bit misleading.

In general, you want to name the class just "webapp-img", because that's the element this is associated with.
You don't want to name it wrt what it's doing.
Another thing is, you're doing more than just the resizing here. There's the background setting as well.

File: portfolio/src/main/webapp/index.html [hunk 11]
```diff
@@ -63,47 +60,65 @@ <h2 class="text-white mb-4">About Me</h2>
                 </div>
             </div>
         </section>
+        <!-- TODO: change descriptions to bullet points -->
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/index.html:63]: Same note about this TODO as above.

Also end with "."

File: portfolio/src/main/webapp/index.html [hunk 12]
```diff
@@ -63,47 +60,65 @@ <h2 class="text-white mb-4">About Me</h2>
...
                     <div class="col-lg-6">
                         <div class="bg-black text-center h-100 project">
                             <div class="d-flex h-100">
                                 <div class="project-text w-100 my-auto text-center text-lg-left">
-                                    <h4 class="text-white">Misty</h4>
-                                    <p class="mb-0 text-white-50">An example of where you can put an image of a project, or anything else, along with a description.</p>
+                                    <h4 class="text-white">Full Stack @ AiCure</h4>
+                                    <p class="mb-0 text-white-50">Built product features from the database all the way to UI, Implemented e2e UI automation testing framework, Built AWS API gateway integration testing framework, 
+                                        Developed ETL tools to support our application testing environment data setup and migration
+                                    </p>
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/index.html:100]:
```suggestion
                                      Developed ETL tools to support our application testing environment data setup and migration
                                    </p>
```

Comment (hayleejane3) [reply portfolio/src/main/webapp/index.html:100]: I think the send lines should be indented more toward the left: https://google.github.io/styleguide/htmlcssguide.html#Separation_of_Concerns

File: portfolio/src/main/webapp/index.html [hunk 13]
```diff
@@ -63,47 +60,65 @@ <h2 class="text-white mb-4">About Me</h2>
...
                         <div class="bg-black text-center h-100 project">
                             <div class="d-flex h-100">
                                 <div class="project-text w-100 my-auto text-center text-lg-right">
-                                    <h4 class="text-white">Mountains</h4>
-                                    <p class="mb-0 text-white-50">Another example of a project with its respective description. These sections work well responsively as well, try this theme on a small screen!</p>
+                                    <h4 class="text-white">CodaLab Competition: Sarcasm Detection</h4>
+                                    <p class="mb-0 text-white-50">Implemented word, sentence-level embeddings, and attention LSTM models, ELMo embeddings with different classifiers using Python (keras, tensorflow) to detect sarcasm in tweets and Reddit posts, obtaining 8th place with a f1 score of 0.756
+                                        Paper will be published at the FigLang 2020 which includes literature about different word embeddings, transformers (BERT, ELMo), and additional classifiers that further improved performance</p>
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/index.html:121]: Same indentation here as above.

File: portfolio/src/main/webapp/index.html [hunk 14]
```diff
@@ -26,7 +26,7 @@
                     <ul class="navbar-nav ml-auto">
                         <li class="nav-item"><a class="nav-link js-scroll-trigger" href="#about">About</a></li>
                         <li class="nav-item"><a class="nav-link js-scroll-trigger" href="#projects">Portfolio</a></li>
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/index.html:28]: Here and everywhere, we indent by 2 spaces, not 4.

The same for CSS too.

File: portfolio/src/main/webapp/js/customscript.js [hunk 15]
```diff
@@ -26,3 +26,27 @@ function addRandomGreeting() {
   const greetingContainer = document.getElementById('greeting-container');
   greetingContainer.innerText = greeting;
 }
+
+// adds Hello Natalie! to the page
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/customscript.js:30]: Some convention. Same for the comments below.

```suggestion
// Adds "Hello Natalie!" to the page.
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (hayleejane3) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:31]: I would also rename this to something more descriptive. Describe what exactly these messages are for, eg: "welcomeMessages". In your case "<something>CityNames".

File: portfolio/src/main/webapp/js/customscript.js [hunk 16]
```diff
@@ -26,3 +26,27 @@ function addRandomGreeting() {
...
+
+//parse the ArrayList as JSON
+function changeAListJson(){
+  fetch('/data')  // sends a request to /my-data-url
+  .then(response => response.json()) // parses the response as JSON
+  .then((hardcodeMessages) => { // now we can reference the fields in hardcodeMessages!
+  console.log(hardcodeMessages[0]);
+  console.log(hardcodeMessages[1]);
+  console.log(hardcodeMessages[2]);
+  console.log('Adding hardcode messages to dom: ' + hardcodeMessages);
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/customscript.js:48]: Update the log statement and the id name to represent the new relevant version of the "hardcodeMessages"

**Comment by hayleejane3** [comment]
Something I noticed: Since the image of you turns colored when I hover over it, I expect it to be clickable.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (ntarn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:34]: This ArrayList was just from following the tutorial, but will do so for future arrays! :)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (ntarn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:49]: Modified the comment!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 17]
```diff
@@ -19,14 +19,35 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
+import com.google.gson.Gson;
```
Comment (lawrencezjh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:24]: wrong import order

Comment (hayleejane3) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:24]: @ntarn We use alphabetical order.

File: portfolio/src/main/webapp/css/customstyles.css [hunk 18]
```diff
@@ -69,11 +69,22 @@ h1{
...
+    width: 1.5rem;
+    background: rgba(100, 161, 157, 0.1);
+    border-radius: 100%;
+    /* line-height: 3rem; */
+    color: rgba(255, 255, 255, 0.3);
+    
+}
+
 .project-icons a {
   text-align: center;
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/css/customstyles.css:84]: Use alphabetical  order for all these too. (Here and elsewhere too)

File: portfolio/src/main/webapp/index.html [hunk 19]
```diff
@@ -112,66 +127,87 @@ <h4 class="text-white">Mountains</h4>
...
                             </div>
                         </div>
                     </div>
                     <div class="col-md-4 mb-3 mb-md-0">
                         <div class="card py-4 h-100">
                             <div class="card-body text-center">
-                                <i class="fas fa-envelope text-primary mb-2"></i>
-                                <h4 class="text-uppercase m-0">Email</h4>
+                                <i class="fas fa-school text-primary mb-2"></i>
+                                    <!-- TODO: Resize <img class="" src="assets/img/browser.svg" alt="" /> -->
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/index.html:150]: Assign TODO to self

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (hayleejane3) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: Could you change it to an array  here too?

Or rather, a const array. 

Since you're not going to be modifying the content.
Modification usually the main reason behind using an ArrayList.

File: portfolio/src/main/webapp/js/customscript.js (hunk 16 above)
Comment (ntarn) [reply portfolio/src/main/webapp/js/customscript.js:48]: This was from the tutorial, but will do so for later objects!

**Comment by ntarn** [comment]
> Something I noticed: Since the image of you turns colored when I hover over it, I expect it to be clickable.

Addressed in private message,

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 20]
```diff
@@ -14,19 +14,39 @@
...
+  }
+
+  /**
+   * Converts an ArrayList into a JSON string using the Gson library. Note: We first added
+   * the Gson library dependency to pom.xml.
+   */
+  private String convertToJsonUsingGson(ArrayList aList) {
+    Gson gson = new Gson();
+    String json = gson.toJson(aList);
+    return json;
```
Comment (lawrencezjh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:50]: this helper function itself is pretty boring. Can you just remove it and just simply do:
new Gson().toJson(blahblahblah)
where you need to use this?

File: portfolio/src/main/webapp/js/customscript.js [hunk 21]
```diff
@@ -26,3 +26,27 @@ function addRandomGreeting() {
...
+//Parse the ArrayList as JSON.
+function changeAListJson(){
+  fetch('/data')  // Sends a request to /my-data-url.
+  .then(response => response.json()) // Parses the response as JSON.
+  .then((hardcodeMessages) => { // Now we can reference the fields in hardcodeMessages.
+  console.log(hardcodeMessages[0]);
+  console.log(hardcodeMessages[1]);
+  console.log(hardcodeMessages[2]);
+  console.log('Adding hardcode messages to dom: ' + hardcodeMessages);
+  document.getElementById('hardcode-messages').innerHTML = hardcodeMessages;
```
Comment (lawrencezjh) [inline portfolio/src/main/webapp/js/customscript.js:49]: I believe you need to indent these lines within the block.
.then((hardcodeMessages) -> {
  blahblahblah...
}

File: portfolio/src/main/webapp/index.html [hunk 22]
```diff
@@ -112,66 +127,87 @@ <h4 class="text-white">Mountains</h4>
...
+            <div class="container">
+                <div class="row">
+                    <div class="col-md-10 col-lg-8 mx-auto text-center">
+                        <i class="far fa-paper-plane fa-2x mb-2 text-white"></i>
+                        <h2 class="text-white mb-5">Subscribe to receive updates!</h2>
+                        <form class="form-inline d-flex"><input class="form-control flex-fill mr-0 mr-sm-2 mb-3 mb-sm-0" id="inputEmail" type="email" placeholder="Enter email address..." /><button class="btn btn-primary mx-auto" type="submit">Subscribe</button></form>
+                    </div>
+                </div>
+            </div>
+        </section> -->
```
Comment (lawrencezjh) [inline portfolio/src/main/webapp/index.html:201]: Remove commented out code

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 23]
```diff
@@ -14,19 +14,39 @@
 
 package com.google.sps.servlets;
 
+import com.google.gson.Gson;
 import java.io.IOException;
+import java.util.ArrayList;
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:19]: Shouldn't be needed if you switch to using the const array

File: portfolio/src/main/webapp/css/customstyles.css [hunk 24]
```diff
@@ -49,34 +49,44 @@ h1{
...
 }
 
+.project-icons #projecticonborder{
+    background: rgba(100, 161, 157, 0.1);
+    border-radius: 100%;
+    color: rgba(255, 255, 255, 0.3);
+    height: 1.5rem;
+    text-align: center;
+    width: 1.5rem;
+
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/css/customstyles.css:79]: Extra new line

File: portfolio/src/main/webapp/index.html [hunk 25]
```diff
@@ -63,47 +60,65 @@ <h2 class="text-white mb-4">About Me</h2>
...
                         <div class="bg-black text-center h-100 project">
                             <div class="d-flex h-100">
                                 <div class="project-text w-100 my-auto text-center text-lg-right">
-                                    <h4 class="text-white">Mountains</h4>
-                                    <p class="mb-0 text-white-50">Another example of a project with its respective description. These sections work well responsively as well, try this theme on a small screen!</p>
+                                    <h4 class="text-white">CodaLab Competition: Sarcasm Detection</h4>
+                                    <p class="mb-0 text-white-50">Implemented word, sentence-level embeddings, and attention LSTM models, ELMo embeddings with different classifiers using Python (keras, tensorflow) to detect sarcasm in tweets and Reddit posts, obtaining 8th place with a f1 score of 0.756
+                                       Paper will be published at the FigLang 2020 which includes literature about different word embeddings, transformers (BERT, ELMo), and additional classifiers that further improved performance</p>
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/index.html:121]: This line should be indented a little more to the left: https://google.github.io/styleguide/htmlcssguide.html#Separation_of_Concerns

File: portfolio/src/main/webapp/index.html [hunk 26]
```diff
@@ -112,66 +127,87 @@ <h4 class="text-white">Mountains</h4>
...
                             </div>
                         </div>
                     </div>
                     <div class="col-md-4 mb-3 mb-md-0">
                         <div class="card py-4 h-100">
                             <div class="card-body text-center">
-                                <i class="fas fa-envelope text-primary mb-2"></i>
-                                <h4 class="text-uppercase m-0">Email</h4>
+                                <i class="fas fa-school text-primary mb-2"></i>
+                                    <!-- TODO(ntarn): Resize <img class="" src="assets/img/browser.svg" alt="" /> -->
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/index.html:150]: Indent comment to be in line with the <i> since it's at the same level.

File: portfolio/src/main/webapp/index.html [hunk 27]
```diff
@@ -112,66 +127,87 @@ <h4 class="text-white">Mountains</h4>
...
                     </div>
                 </div>
+            </div>
+        </section>
+        <!-- Tutorial Section-->
+        <section class="bg-black" id="contact">
+            <div id="container">
+                <h1>Testing</h1>
+                <p>Click here to see a Hello Natalie:</p>
+                 <button onclick="getHelloNameUsingAsyncAwait()">Fetch a Hello Name!</button>
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/index.html:177]:
```suggestion
                <button onclick="getHelloNameUsingAsyncAwait()">Fetch a Hello Name!</button>
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (hayleejane3) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: My bad, I thought this was JS.

I would recommend using an ImmutableList instead since you're not changing the values
Eg: https://guava.dev/releases/23.0/api/docs/com/google/common/collect/ImmutableList.html

File: portfolio/src/main/webapp/js/customscript.js (hunk 16 above)
Comment (hayleejane3) [reply portfolio/src/main/webapp/js/customscript.js:48]: I guess my point was that you're customizing this site to be your own, so you can name things to be more appropriate to what purpose they serve. The tutorials would use more literal names to describe what they're trying to teach you, i.e. teach you about hardcoding messages. But since you know that you're hardcoding messages and these messages serve 'x' purpose, you could say something like xMessages or x<something>. 
Basically relevant to what it is, since you can have more than one instance of hardcoded messages. So you want to be as specific about each one as possible.

**Comment by ntarn** [comment]
I forgot to delete the //Arraylist comment in DataServlet.java, but will do that in the other branch

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 28]
```diff
@@ -29,24 +30,44 @@
  */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+	ImmutableList<String> immutableList =  ImmutableList.of("blob1", "For", "blob"); 
+  //ArrayList<String> hardcodeMessages = new ArrayList<String>(Arrays.asList("Buenos Aires", "Cordoba", "La Plata"));
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:34]: Remove unused code.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 29]
```diff
@@ -29,24 +30,44 @@
  */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+	ImmutableList<String> immutableList =  ImmutableList.of("blob1", "For", "blob");
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: If you plan to remove this and replace it with the whole commenting feature, you can just take it out and put if back if you actually want it back.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 30]
```diff
@@ -29,24 +30,44 @@
...
+    // Respond with the result.
+    comments.add(text);
+    response.setContentType("text/html;");
+    for(int commentIndex =0; commentIndex < comments.size(); commentIndex++){
+      response.getWriter().println(comments.get(commentIndex));
+    }
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/index.html");
+
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:59]: Extra line.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 31]
```diff
@@ -29,24 +30,44 @@
...
+    // Redirect back to the HTML page.
+    response.sendRedirect("/index.html");
+    
+  }
+
   /**
-   * Converts an ArrayList into a JSON string using the Gson library. Note: We first added
-   * the Gson library dependency to pom.xml.
+   * @return the request parameter, or the default value if the parameter
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:63]: Add comment above also hat says "Returns...". That doesn't need the details on the default case and such, just the general behavior.
It does feel redundant, but convention.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 32]
```diff
@@ -29,24 +30,44 @@
...
+   *         was not specified by the client
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:64]:
```suggestion
   *         was not specified by the client.
```

File: portfolio/src/main/webapp/css/customstyles.css [hunk 33]
```diff
@@ -49,20 +49,20 @@ h1{
     display: inline;
 } */
 
-/* Changes the grayscale profile picture to colored when hovered over. */
+/* Changes the grayscale profile picture to colored when hovered over.*/
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/css/customstyles.css:52]: I would move this comment to the exact line(s) that do this. 

That's because this entire block in itself does more than just the grayscale. You're applying general styles to the entire element. 

The comment would be outside if the CSS rules in the block were all exclusively for the grayscaling. In which case, the class name would also be something like .grayscale{...} and then you could reuse it in multiple places to grayscale a lot of things.

File: portfolio/src/main/webapp/js/customscript.js [hunk 34]
```diff
@@ -37,16 +37,27 @@ async function getHelloNameUsingAsyncAwait() {
...
-  fetch('/data')  // Sends a request to /my-data-url.
+  fetch('/data')  // Sends a request to /data .
   .then(response => response.json()) // Parses the response as JSON.
-  .then((hardcodeMessages) => { // Now we can reference the fields in hardcodeMessages.
-  console.log(hardcodeMessages[0]);
-  console.log(hardcodeMessages[1]);
-  console.log(hardcodeMessages[2]);
-  console.log('Adding hardcode messages to dom: ' + hardcodeMessages);
-  document.getElementById('hardcode-messages').innerHTML = hardcodeMessages;
+  .then((testGetMethodMessages) => { // Now we can reference the fields in hardcodeMessages.
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/customscript.js:44]: Update comment?

File: portfolio/src/main/webapp/js/customscript.js [hunk 35]
```diff
@@ -37,16 +37,27 @@ async function getHelloNameUsingAsyncAwait() {
...
+    console.log('Adding hardcode messages to dom: ' + testGetMethodMessages);
+    document.getElementById('hardcode-messages').innerHTML = testGetMethodMessages;
+});
+}
+
+// Parse the ArrayList comments as JSON.
+function printComments(){
+  fetch('/data')  // Sends a request to /data .
+  .then(response => response.json()) // Parses the response as JSON.
+  .then((comments) => { // Now we can reference the fields in hardcodeMessages.
```
Comment (hayleejane3) [inline portfolio/src/main/webapp/js/customscript.js:57]: Ditto here.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 36]
```diff
@@ -14,19 +14,57 @@
...
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String text = getParameter(request, "text-input", "");
+
+    // Respond with the result.
+    comments.add(text);
     response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    for(int commentIndex =0; commentIndex < comments.size(); commentIndex++){
```
Comment (lawrencezjh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:50]: space between '=' and '0'

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 37]
```diff
@@ -14,19 +14,57 @@
...
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String text = getParameter(request, "text-input", "");
+
+    // Respond with the result.
+    comments.add(text);
```
Comment (lawrencezjh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:48]: I don't know if I'm missing any context here, but I've never seen the definition of 'comments' within this file. Where is this var defined?

File: portfolio/src/main/webapp/js/customscript.js [hunk 38]
```diff
@@ -26,3 +26,24 @@ function addRandomGreeting() {
...
+
+// Parse the ArrayList comments as JSON.
+function printComments(){
+  fetch('/data')  // Sends a request to /data .
+  .then(response => response.json()) // Parses the response as JSON.
+  .then((comments) => { // Now we can reference the fields in comments.
+    console.log(comments);
+    console.log('Adding comments to dom: ' + comments);
+    document.getElementById('print-comments').innerHTML = comments;
+});
```
Comment (lawrencezjh) [inline portfolio/src/main/webapp/js/customscript.js:48]: this '})' should be aligned with the previous '.then', so it should be indented by 2 spaces.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 36 above)
Comment (lawrencezjh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:50]: also space between 'for' and '('

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 37 above)
Comment (ntarn) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:48]: The comments got removed later on in another branch because it was replaced with a different data structure and logic, but I removed it here to fix a merge conflict that happened with conflicting variable names.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 39]
```diff
@@ -50,18 +46,18 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
       response.getWriter().println(comments.get(commentIndex));
     }
 
     // Redirect back to the HTML page.
     response.sendRedirect("/index.html");
-    
   }
 
   /**
+   * Returns the request parameter.
```
Comment (hayleejane3) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:58]: Add an empty line before the @return line

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 37 above)
Comment (lawrencezjh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:47]: I still don't under quite understand: does that mean when you merge this into master, this Java file won't compile at all? Because variable 'comments' is clearly never defined within this file.

File: portfolio/src/main/webapp/js/customscript.js [hunk 40]
```diff
@@ -26,3 +26,24 @@ function addRandomGreeting() {
...
+  console.log('Fetching Hello Name!');
+  const response = await fetch('/data');
+  console.log('Handling the response.');
+  const name = await response.text();
+  console.log('Adding quote to dom: ' + name);
+  document.getElementById('quote-container').innerHTML = name;
+}
+
+// Parse the ArrayList comments as JSON.
+function printComments(){
```
Comment (lawrencezjh) [inline portfolio/src/main/webapp/js/customscript.js:41]: space between '()' and '{'
